More H2 Queries

paulcalvano opened the pull request:
Closes #101

rviscomi commented on sql/2019/20_HTTP_2/20_06.sql:
```diff
@@ -0,0 +1,30 @@
+
+SELECT 
+  _TABLE_SUFFIX AS client,
+  JSON_EXTRACT_SCALAR(payload, "$._is_base_page") AS is_base_page,  
```
Ditto previous comments. Should `is_base_page` be a field in the output or a condition in the WHERE clause? Applies to the rest of these queries as well.

paulcalvano commented:
Added 20.16, which includes alt-svc headers at @bazzadp's request.

paulcalvano commented on sql/2019/20_HTTP_2/20_04.sql:
```diff
@@ -0,0 +1,29 @@
+SELECT 
+  _TABLE_SUFFIX AS client,
+  JSON_EXTRACT_SCALAR(payload, "$._is_base_page") AS is_base_page,  
+  COUNT(*) AS num_requests
```
Seeing a discrepancy in the results between the query in my PR and the one you shared above.   In the screenshot below, the query on the left processes the  httparchive.requests tables and the one on the right processes the httparchive.almanac.requests table.   

The one on the left shows 1890586 mobile HTTP/2 pages.  The one on the right shows 1996270 mobile HTTP/2 pages . I don't see any reason for the discrepency in the queries.  Any ideas? 

![image](https://user-images.githubusercontent.com/7459458/65783581-13dc8b80-e11e-11e9-8089-f055806eb7e1.png)

rviscomi replied:
Great catch. Seems like a difference in requests marked `_is_base_page` and `firstHtml`. The former is a WPT annotation while the latter is determined in the HA pipeline.

Looking at the number of base pages:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  COUNT(0)
FROM
  `httparchive.requests.2019_07_01_*`
WHERE
  JSON_EXTRACT_SCALAR(payload, '$._is_base_page') = 'true'
GROUP BY
  client
```

There are 5,297,105 mobile and 4,371,570 desktop base pages. Compared to the right side totals in your screenshot, which are 5,558,214 and 4,550,580 respectively.

The actual number of pages in the `summary_pages` dataset for the 2019_07 crawl is:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  COUNT(0)
FROM
  `httparchive.summary_pages.2019_07_01_*`
GROUP BY
  client
```

5,297,442 and 4,371,973 respectively. So much closer to the base page numbers. I'm not even sure how we get ~200k _more_ `firstHtml` requests than there are pages. 😖 

```sql
#standardSQL
SELECT
  client,
  page,
  COUNT(0) AS firstHtml
FROM
  `httparchive.almanac.requests`
WHERE
  firstHtml
GROUP BY
  client,
  page
HAVING
  firstHtml > 1
ORDER BY
  firstHtml DESC
```

This reveals MANY pages in which there are more than 1 `firstHtml` value. Some with 400+. On [this example page](http://httparchive.webpagetest.org/result/190713_Dx89_G5WS/1/details/#waterfall_view_step1) the page makes over 1000 requests, many to `/index.php` and `/`. Maybe the HA pipeline saw the first request for `/`, marked it as `firstHtml`, then marked all subsequent requests for that URL also as `firstHtml`?

In any case, this bug affects all other queries that depend on `firstHtml`. I think the best path forward for now is to open a bug to get this fixed in the pipeline for future crawls, overwrite all `firstHtml` values in the `almanac` dataset to match `_is_base_page`, rerun all queries that include it, and update all results and written analyses that depend on those queries.

Thanks for pointing this out Paul!

rviscomi replied:
I was about to file the bug and was querying the `summary_requests` dataset directly to eliminate any `almanac` dataset shenanigans, but it turns out that there are `almanac` dataset shenanigans!

The last query I shared in the previous comment produced hundreds of thousands of pages with more than 1 `firstHtml` request. It should have produced 0. This equivalent query of the `summary_requests` dataset _does_ produce 0 results:

```sql
#standardSQL
SELECT
  _TABLE_SUFFIX AS client,
  pageid,
  COUNT(0) AS html
FROM
  `httparchive.summary_requests.2019_07_01_*`
WHERE
  firstHtml
GROUP BY
  client,
  pageid
HAVING
  html > 1
ORDER BY
  html DESC
```

Now I think I know what the bug is. When I generated the almanac dataset, if the summary request had `firstHtml` all requests for that page/url were also annotated as `firstHtml`. I should be able to fix this, but we still need to rerun all the queries.

rviscomi replied:
@paulcalvano for now let's change these queries to use `firstHtml` solely for consistency. At least it'll have the same bug as all the other metrics and we can fix them all in one swoop.

rviscomi commented:
I've checked off all remaining metrics in #101 that are covered by this PR. Once the queries are updated to use `firstHtml` rather than `is_base_page` this should be ready for a final review to turn over to @bazzadp for writing.

bazzadp commented:
@paulcalvano Can we change 20.16 to SUM rather than AVG?

```
  ROUND(SUM(num_requests),2) AS num_pushed_requests,
  ROUND(SUM(kb_transfered),2) AS num_kb_transfered
```

Also is there a way to group these into the below categories? I've done a vlookup in the results spreadsheet, but not sure if there is a better way for future years? And categorising by these categories is why I want the sum, rather than the total. I can approximate this (pretty well it turns out!) but multiplying the num pages by average but would be better to have the raw stats. I got VERY different results when I just looked at the averages...

- Video
- --
- Text
- PDF
- JavaScript
- Image
- Font
- Data
- CSS

paulcalvano commented:
Hey Barry.  Sure, I'm updating the queries per Rick's comment earlier (ie, switching to the almanac table and using firstHtml instead of is_base_page). I'm sorry that I'm so behind on this.  

Will add a note to each sheet indicating whether the data is updated to use the almanac table.

bazzadp commented:
Can I also get HTTPS added to the upgrade header spreadsheet?

paulcalvano commented:
> Can I also get HTTPS added to the upgrade header spreadsheet?

Which query is that?

paulcalvano commented:
I updated all the data in the sheet for the queries I updated above.  Will have the rest of these done later today.   THe numbers are very close - so the discrepancy I noted above does not appear to be too impactful and shouldn't affect the analysis.

bazzadp commented:
>> Can I also get HTTPS added to the upgrade header spreadsheet?
> Which query is that?

The 4a, 5a and 6a sheet. Find that one most useful btw so ignore the individual queries as only going to look at this combined one.

paulcalvano commented:
All queries have been updated to use the almanac tables except for 20.15.  There doesn't appear to be a summary_pages almanac table, so I left that one as is.   I've updated all the data in the sheets for the queries I updated.   In most cases, it resulted in only slight changes in the numbers. 

For 20.11, I included both tables because the results were very different between the two queries.    I trust the 2019_07_01 table results more FWIW - so we might just want to keep the only query for that one.   @rviscomi  wdyt?

paulcalvano commented:
Just added the https indicators to two queries @bazzadp .  I didn't want to mess up your pivot tables, so instead of updating those in the sheet I added two extra sheets at the end with the results.

@rviscomi - this is ready for a final review.
